src/model.jl

```
@@ -5,5 +5,9 @@ Return an instance representing a model.
 
 The return-type depends on its usage in [`compute_tempered_logdensities`](@ref).
 """
-function make_tempered_model end
+make_tempered_model(sampler, model, beta) = make_tempered_model(model, beta)
+make_tempered_model(model, beta) = TemperedLogDensityProblem(model, beta)
```

Do we want all cases where `make_tempered_model` is called with two args to result in a `TemperedLogDensityProblem`? I know Turing etc. needs the sampler but maybe we want to avoid such a strict prescription. Is there a way we can specify the list of model types that support LogDensityProblems in the function defn instead? Granted that would mean we would have to edit each time we discover a new one (not sure if many other packages use it). See my bigger comment for context

---

So the `sampler` argument wasn't there to start with because of Turing; it's there so the sampler itself can change how the tempered model is constructor. For example, should the tempering be on the entire density or should it be on just the likelihood? This IMO should just go in the sampler. That's one use-case.

>  Is there a way we can specify the list of model types that support LogDensityProblems in the function defn instead? Granted that would mean we would have to edit each time we discover a new one (not sure if many other packages use it). See my bigger comment for context

This is only the _default_ implementation, while before this method just wasn't implemented. So how difficult  it is for users to implement for a new model is unaffected, right?

I should throw an error though in the case where it doesn't implement the LogDensityProblems.jl interface :+1:

---

Ah sorry, you're right, we don't need the sampler for Turing, I was thinking of something else.

In that case I think I see what you are getting at, but not 100%. Are you saying that the sampler should / may contain configuration that can be propagated through `make_tempered_model` to determine how the `model` is tempered? I.e. for `AdvancedHMC` we could implement it like this but somehow allow the sampler to enable / disable tempering the `Joint` vs just the likelihood?

```julia
struct Joint{Tℓprior, Tℓll} <: Function
    ℓprior      :: Tℓprior
    ℓlikelihood :: Tℓll
end

function (joint::Joint)(θ)
    return joint.ℓprior(θ) .+ joint.ℓlikelihood(θ)
end


struct TemperedJoint{Tℓprior, Tℓll, T<:Real} <: Function
    ℓprior      :: Tℓprior
    ℓlikelihood :: Tℓll
    β           :: T
end

function (tj::TemperedJoint)(θ)
    return tj.ℓprior(θ) .+ (tj.ℓlikelihood(θ) .* tj.β)
end


function MCMCTempering.make_tempered_model(
    model::DifferentiableDensityModel,
    β::Real
)
    ℓπ_β = TemperedJoint(model.ℓπ.ℓprior, model.ℓπ.ℓlikelihood, β)
    ∂ℓπ∂θ_β = TemperedJoint(model.∂ℓπ∂θ.ℓprior, model.∂ℓπ∂θ.ℓlikelihood, β)
    model_β = DifferentiableDensityModel(ℓπ_β, ∂ℓπ∂θ_β)
    return model_β
end
```



Fair enough on this being a default implementation, I suppose it is a subjective debate on how easy it is for someone to extend. On one level it may make someone feel they *have* to use `LogDensityProblems` to implement tempering / we are encouraging that. On the other hand I agree it doesn't have much tangible impact on difficulty. It is more just that if we are providing implementations of these parts to facilitate tempering perhaps they belong in the packages they are specific too, rather than this "general" tempering package?

Agree on the error as a solution for this approach though.

---

> Fair enough on this being a default implementation, I suppose it is a subjective debate on how easy it is for someone to extend.

The way I've set it up there are basically levels to it (I was about to add it to the docs but realized we don't have any; so probably should get that started):
1. Easy way (but not the most efficient): implement `getparms(transition)` to return a `Vector` from your `transition` type and, then if you
   1. implement LogDensityProblems.jl ⟹ you're done,
   1. otherwise: implement `MCMCTempering.logdensity(model, x)` and `MCMCTempering.make_tempered_model(model, beta)` and you're done.
1. Continuing from the above, if you want more performance:
   1. implement `MCMCTempering.compute_tempered_logdensities` explicitly for the `model` and `transition` types of interest so that you can make use of the _cached_ log-joint computations usually present in a `transition`.

This isn't _that_ complicated, is it? :confused:

> On one level it may make someone feel they have to use LogDensityProblems to implement tempering / we are encouraging that. 

It's worth pointing out that this is the default in AbstractMCMC.jl now (as in, if the provided model doesn't have it's own definition of something like `step` it will assume it's an implementation of LogDensityProblems.jl + check with descriptive error if not), and the reason for this is because it means less different `AbstractModel` types to deal with in the ecosystem, e.g. both AdvancedHMC.jl and AdvandedMH.jl can use the LogDensityProblems.jl interface at no cost :shrug: 

Worth pointing out that the interface is just:
- `logdensity(f, x)`
- `dimension(f)`: return dimensionality of the expected input
- `compatibilities`: return which order of differentiation we have access to

With all that being said, if you don't feel comfortable encouraging it (I do becuase it automatically means that it's compatible with many other sampler packages), we can just remove the mentioning of LogDensityProblems.jl from the error-messages:) Then there will be not mentioning/encouragement of usage of LogDensityProblems.jl.

> It is more just that if we are providing implementations of these parts to facilitate tempering perhaps they belong in the packages they are specific too, rather than this "general" tempering package?

But that's what we are though, no? It's just a question of: should we provide a default-implementation of these general mechanism using a _super_ low-cost interface such as LogDensityProblems.jl, which _immediately_ gives us compat with many of the existing sampler packages (without having to depend on them, mind you)? IMO, yes. The result is that you only have to implement a single method (`getparams(transition)`) to get up and running, and that's awesome.

_And_, maybe most importantly, it allows samplers to be compat with MCMCTempering _without explicit dependency_. This such an important thing because even though most implementations of samplers will be happy for their sampler to be used in tempering, they'll be less happy about taking on the responsibility for maintaining that compatibility (in particular if they don't own the tempering package). Now they can just implement this super-simple, super-barebones (and very stable) minimal interface of LogDensityProblems.jl and their users can suddenly do tempering:)

---

I updated the README (https://github.com/TuringLang/MCMCTempering.jl/tree/torfjelde/logdensity-support#tutorial-supporting-mcmctempering) now.

---

Sure yeah, I just wanted to raise the topic as I feel its important to reflect on the decisions. Reading through your thinking I agree with it all. I am happy to go with this. Didn't realise quite how lightweight the framework was too, which yeah does make a lot of sense alongside the existing tie in with AbstractMCMC. Thanks for the in depth explanation. Agree on the docs, this stuff should go in there in some capacity for sure!

---

No no no, it's def good to discuss these decisions! So didn't mean to come off as "combative" or dismissive; if that was case, I'm sorry. But awesome:)